src/lib/core/backtrace.h

```diff
@@ -1,64 +1,86 @@
-#define TARANTOOL_LIB_CORE_BACKTRACE_H_INCLUDED
 /*
- * Copyright 2010-2016, Tarantool AUTHORS, please see AUTHORS file.
+ * SPDX-License-Identifier: BSD-2-Clause
  *
- * Redistribution and use in source and binary forms, with or
- * without modification, are permitted provided that the following
- * conditions are met:
- *
- * 1. Redistributions of source code must retain the above
- *    copyright notice, this list of conditions and the
- *    following disclaimer.
- *
- * 2. Redistributions in binary form must reproduce the above
- *    copyright notice, this list of conditions and the following
- *    disclaimer in the documentation and/or other materials
- *    provided with the distribution.
- *
- * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
- * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
- * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
- * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
- * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
- * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
- * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
- * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
- * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
- * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
- * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
- * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
- * SUCH DAMAGE.
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
  */
+
+#pragma once
+
 #include "trivia/config.h"
-#include <stddef.h>
+#include "trivia/util.h"
 
-#if defined(__cplusplus)
-extern "C" {
-#endif /* defined(__cplusplus) */
+#include "libunwind.h"
 
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
 #ifdef ENABLE_BACKTRACE
-#include <coro.h>
+/*
+ * Format for printing C/C++ frames.
+ */
+#define C_FRAME_STR_FMT "#%-2d %p in %s+%zu"
 
-char *
-backtrace(char *start, size_t size);
+enum {
+	/* Maximal number of frames collected. */
+	BACKTRACE_FRAME_COUNT_MAX = 64,
+};
 
-void print_backtrace(void);
+/*
+ * C/C++ frame information sufficient for further resolving of function name
+ * and offset.
+ */
+struct backtrace_frame {
+	/* Value of IP register. */
+	void *ip;
+};
 
-typedef int (backtrace_cb)(int frameno, void *frameret,
-                           const char *func, size_t offset, void *cb_ctx);
+/*
+ * Collection of C/C++ frames.
+ */
+struct backtrace {
+	/* Number of frames collected. */
+	int frame_count;
+	/* Array of frames to be further resolved. */
+	struct backtrace_frame frames[BACKTRACE_FRAME_COUNT_MAX];
+};
 
+struct fiber;
 
-void
-backtrace_foreach(backtrace_cb cb, coro_context *coro_ctx, void *cb_ctx);
+/*
+ * Collect call stack of `fiber` (only C/C++ frames) to `bt`.
+ *
+ * `skip_frames` determines the number of frames skipped, starting from the
+ * frame of `backtrace_collect`. It is expected to have a non-negative value.
```

Comment:
Please add an example for clarity. Something like this:

```
For example, if skip_frames is 0, then the function will include backtrace_collect
and every calling function into the stack. If it's 1, it will skip backtrace_collect.
```

Reply:
Fixed: supplied an example.